decoder.go

```diff
@@ -816,31 +816,31 @@ func (d *Decoder) DecodeNested(m interface{}) error {
 	if d.offset >= len(d.p) {
 		return io.ErrUnexpectedEOF
 	}
-	l, n, err := decodeVarint(d.p[d.offset:])
+	l, n, err := DecodeVarint(d.p[d.offset:])
 	if err != nil {
 		return err
 	}
 	if n == 0 {
 		return ErrInvalidVarintData
 	}
-	if l == 0 {
-		d.offset += n
-		return nil
+	nb := int(l)
+	if nb < 0 {
+		return fmt.Errorf("csproto: bad byte length %d", nb)
 	}
-	if d.offset+n+int(l) > len(d.p) {
+	if d.offset+n+nb > len(d.p) {
 		return io.ErrUnexpectedEOF
```

Comment:
curiosity: would it be valuable to wrap the unexpected EOF error with any additional context?

Reply:
Possibly.  I chose to return a "plain" `io.ErrUnexpectedEOF` because that's what the underlying runtimes do for the same scenario.